src/main/java/com/redhat/devtools/lsp4ij/features/formatting/LSPServerSideOnTypeFormattingHelper.java

```diff
@@ -0,0 +1,151 @@
+        Position position = LSPIJUtils.toPosition(offset, document);
+        DocumentOnTypeFormattingParams onTypeFormattingParams = new DocumentOnTypeFormattingParams(
+                LSPIJUtils.toTextDocumentIdentifier(file.getVirtualFile()),
+                new FormattingOptions(LSPIJUtils.getTabSize(editor), LSPIJUtils.isInsertSpaces(editor)),
+                position,
+                charTypedAsString
+        );
+        CompletableFuture<List<TextEdit>> onTypeFormattingFutures = onTypeFormattingSupport.onTypeFormatting(onTypeFormattingParams);
+        try {
+            waitUntilDone(onTypeFormattingFutures, file);
+        } catch (ProcessCanceledException e) {
+            //Since 2024.2 ProcessCanceledException extends CancellationException so we can't use multicatch to keep backward compatibility
+            //TODO delete block when minimum required version is 2024.2
+            onTypeFormattingSupport.cancel();
+            return false;
+        } catch (CancellationException e) {
+            onTypeFormattingSupport.cancel();
+            return false;
+        } catch (ExecutionException e) {
+            LOGGER.error("Error while consuming LSP 'textDocument/onTypeFormatting' request", e);
+            return false;
+        }
+
+        if (!isDoneNormally(onTypeFormattingFutures)) {
+            return false;
+        }
+
+        // If on-type formatting resulted in text edits, apply them
+        List<TextEdit> textEdits = onTypeFormattingFutures.getNow(Collections.emptyList());
+        if (ContainerUtil.isEmpty(textEdits)) {
+            return false;
+        }
+
+        LSPIJUtils.applyEdits(editor, document, textEdits);
+        return true;
+    }
+
+    /**
+     * Returns the LSP on-type formatting information for the provided file if supported. This includes the on-type
+     * formatting options and the on-type formatting support.
+     *
+     * @param file the PSI file
+     * @return the LSP on-type formatting information for the file if supported; otherwise null
+     */
+    @Nullable
+    private static Pair<DocumentOnTypeFormattingOptions, LSPOnTypeFormattingSupport> getOnTypeFormattingInfo(@NotNull PsiFile file) {
+        // On-type formatting shouldn't trigger a language server to start
+        Project project = file.getProject();
+        Set<LanguageServerWrapper> startedLanguageServers = LanguageServiceAccessor.getInstance(project).getStartedServers();
+        for (LanguageServerWrapper startedLanguageServer : startedLanguageServers) {
+            // TODO: Is there a better way to ask if this started language server supports the file?
+            if (startedLanguageServer.isConnectedTo(LSPIJUtils.toUri(file))) {
+                LSPOnTypeFormattingFeature onTypeFormattingFeature = startedLanguageServer.getClientFeatures().getOnTypeFormattingFeature();
+                if (onTypeFormattingFeature.isEnabled(file) && onTypeFormattingFeature.isSupported(file)) {
```

Comment:
Then we just make sure that the on-type formatting feature is enabled and supported and, if so, we return both the options -- which are needed to check whether or not the typed character is an on-type formatting trigger -- and the support -- which is needed to actually invoke the on-type formatting LSP feature.